cmd/changers.go

```diff
@@ -0,0 +1,105 @@
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/lf-edge/eden/pkg/controller"
+	"github.com/lf-edge/eden/pkg/controller/adam"
+	"github.com/lf-edge/eden/pkg/device"
+	"github.com/lf-edge/eve/api/go/config"
+	"github.com/spf13/viper"
+	"io/ioutil"
+	"os"
+	"strings"
+)
+
+type configChanger interface {
+	getControllerAndDev() (controller.Cloud, *device.Ctx, error)
+	setControllerAndDev(controller.Cloud, *device.Ctx) error
+}
+
+type fileChanger struct {
+	fileConfig string
+}
+
+func (ctx *fileChanger) getControllerAndDev() (controller.Cloud, *device.Ctx, error) {
+	if _, err := os.Lstat(ctx.fileConfig); os.IsNotExist(err) {
+		return nil, nil, err
+	}
+	var ctrl controller.Cloud = &controller.CloudCtx{Controller: &adam.Ctx{}}
+	data, err := ioutil.ReadFile(ctx.fileConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("file reading error: %s", err)
+	}
+	var deviceConfig config.EdgeDevConfig
+	err = json.Unmarshal(data, &deviceConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unmarshal error: %s", err)
+	}
+	dev, err := ctrl.ConfigParse(&deviceConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("configParse error: %s", err)
+	}
+	return ctrl, dev, nil
+}
+
+func (ctx *fileChanger) setControllerAndDev(ctrl controller.Cloud, dev *device.Ctx) error {
+	res, err := ctrl.GetConfigBytes(dev)
+	if err != nil {
+		return fmt.Errorf("GetConfigBytes error: %s", err)
+	}
+	if err = ioutil.WriteFile(ctx.fileConfig, res, 0755); err != nil {
+		return fmt.Errorf("WriteFile error: %s", err)
+	}
+	return nil
+}
+
+type adamChanger struct {
+	adamUrl string
+}
+
+func (ctx *adamChanger) getControllerAndDev() (controller.Cloud, *device.Ctx, error) {
+	ipPort := strings.Split(ctx.adamUrl, ":")
```

Comment:
I don't think parsing of the CLI arguments belongs in the utility function -- it probably belongs to adamChanger context that neeeds to be populated from the actual cmd line function. Lets refactor it that way so that these functions simply operate on well-populated context objects.

Makes sense?